### PR TITLE
feat: Updated Factory Composition and Inheritance

### PR DESCRIPTION
**Pull request by henryuks129**
Factory Composition and Inheritance Feature #40 

Overview
This PR introduces factory composition and inheritance capabilities to Interface-Forge, empowering developers to construct more sophisticated and modular data structures. By extending and composing factories, teams can now reuse base configurations while customizing output for specific contexts—dramatically improving productivity and code maintainability.

Features Added
1. Factory Inheritance (extend method)
- Allows creation of child factories from a base factory
- Supports type-safe overrides and enhancements
- Seamless TypeScript support with generic typing

TS
const BaseUserFactory = new Factory<BaseUser>((factory) => ({
  id: factory.datatype.uuid(),
  createdAt: factory.date.recent(),
}));

const AdminUserFactory = BaseUserFactory.extend<AdminUser>(() => ({
  role: 'admin',
  permissions: ['read', 'write', 'delete'],
}));

2.  Factory Composition (compose method)
* Combine factories and values into one unified output
* Supports batch and static object merges
* Type-safe output with full inference
* Ideal for building relational data structures

TS
const UserWithPostsFactory = UserFactory.compose<UserWithPosts>({
  posts: PostFactory.batch(3),
  status: 'active'
});

Technical Details
Type System Enhancements
* Introduced FactoryComposition<T> for typed object merging
* Improved inference in extend and compose methods
* Enhanced support for nested Record and Array values

Code Additions

* extend<U extends T>(...) and compose<U extends T>(...) added to the Factory class
* New utility types for safe merging and output shaping
* Refactored internal parseValue and helper casting logic

Testing

* Unit tests for extend method
* Unit tests for compose method
* Type safety assertions
* Edge case checks and regression coverage

Documentation Updates
Updated README.md:
  *  Usage examples
  *  Best practices
  *  API method signatures
  *  Inheritance and composition patterns

Checklist
* [x] Added inheritance via extend
* [x] Added composition via compose
* [x] Type-safe implementations
* [x] Updated documentation
* [x] Added relevant tests
* [x] Provided real-world usage examples
* [x] Verified all tests pass
* [x] Type-checked and validated

### CODE REVIEW

**Comment by Goldziher** [comment]
Please fix the validation issues

**Comment by henryuks129** [comment]
> Please fix the validation issues

Apologies for letting the validation issues slip past me. I’ve reviewed the failing checks and worked on them. please review, thank you

File: tests/index.spec.ts [hunk 1]
```diff
@@ -1,4 +1,4 @@
-import { Factory } from './index.js';
+import { Factory } from '../src/index';
```
Comment (Goldziher) [inline tests/index.spec.ts:2]: please return this file to src

File: tsconfig.tsbuildinfo [hunk 2]
```diff
@@ -0,0 +1 @@
...,1],[50,1],[51,1],[8,1],[57,1],[54,1],[55,1],[56,1],[58,1],[9,1],[59,1],[60,1],[61,1],[63,1],[62,1],[64,1],[65,1],[10,1],[66,1],[67,1],[68,1],[11,1],[69,1],[70,1],[71,1],[72,1],[73,1],[1,1],[74,1],[75,1],[12,1],[79,1],[77,1],[81,1],[84,1],[76,1],[80,1],[78,1],[17,1],[16,1],[380,1],[271,209],[281,210],[270,209],[291,211],[262,212],[261,213],[290,214],[284,215],[289,216],[264,217],[278,218],[263,219],[287,220],[259,221],[258,214],[288,222],[260,223],[265,224],[266,1],[269,224],[256,1],[292,225],[282,226],[273,227],[274,228],[276,229],[272,230],[275,231],[285,214],[267,232],[268,233],[277,234],[257,235],[280,226],[279,224],[283,1],[286,236],[552,237],[539,238],[540,237],[538,1],[438,239],[379,240],[354,241],[353,242],[372,243],[370,244],[371,245],[359,246],[360,244],[367,247],[358,248],[363,249],[373,1],[364,250],[369,251],[374,252],[357,253],[365,254],[366,255],[361,256],[368,243],[362,257],[351,242],[350,1],[352,258],[377,1],[376,1],[375,1],[378,259],[551,260],[546,261],[553,262],[533,263],[562,264],[564,265],[556,266],[565,267],[563,268],[557,269],[547,270],[571,271],[572,272],[532,1],[439,1],[516,273],[517,274],[440,275]],"affectedFilesPendingEmit":[516,517,440],"version":"5.8.3"}
```
Comment (Goldziher) [inline tsconfig.tsbuildinfo:1]: please remove this file and add it to the git ignore

File: README.md [hunk 3]
```diff
@@ -246,32 +250,99 @@ console.log(generator.next().value); // 1 (or 2, or 3)
 console.log(generator.next().value); // (different from the previous value)
 ```
 
-## TypeScript Compatibility
+#### `extend`
 
-Interface-Forge is designed to work with TypeScript 5.x and above. It leverages TypeScript's type system to provide type safety and autocompletion for your factories.
+Extends the current factory to create a new factory with additional or overridden properties. This method allows for factory inheritance, a new factory can be built upon an existing one while adding or modifying properties.
 
-## Faker.js Integration
```
Comment (Goldziher) [inline README.md:253]: please return the bits of the readme you removed by mistake

**Comment by Goldziher** [comment]
Nice work!
